code-studio/src/js/levels/dialogHelper.js

```diff
@@ -157,58 +159,68 @@ window.dashboard.dialog = (function () {
    * @param {function(boolean)} onComplete Optional callback function to call when
    *     the server call completes, which receives a boolean indicating whether
    *     the browser will redirect to a new location after it is called.
+   * @param {function(function)} beforeHook Optional callback function to call
+   *     before processResults gets underway. This function must call the completion
+   *     function passed as the parameter to complete the processResults call.
    */
   // TODO(dave): move this logic into appOptions.onAttempt for levels of type
```

Comment:
I'm not sure `/apps` gets this right either.  Maybe we should rethink the events for all app types?  (Which would be beyond the scope of this PR...)